File: doc-examples/src/main/scala/jms/JmsToWebSocket.scala
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <http://www.lightbend.com>
 */

package jms

// #sample
import akka.Done
import akka.http.scaladsl.Http
import akka.http.scaladsl.model._
import akka.http.scaladsl.model.ws.{WebSocketRequest, WebSocketUpgradeResponse}
import akka.stream.alpakka.jms.JmsConsumerSettings
import akka.stream.alpakka.jms.scaladsl.{JmsConsumer, JmsConsumerControl}
import akka.stream.scaladsl.{Flow, Keep, Sink, Source}

import scala.concurrent.Future
// #sample
import playground.{ActiveMqBroker, WebServer}

import scala.concurrent.duration.DurationInt

object JmsToWebSocket extends JmsSampleBase with App {

  ActiveMqBroker.start()
  WebServer.start()

  val connectionFactory = ActiveMqBroker.createConnectionFactory
  enqueue(connectionFactory)("a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k")

  // format: off
  // #sample

  val jmsSource: Source[String, JmsConsumerControl] =
    JmsConsumer.textSource(                                                           // (1)
      JmsConsumerSettings(actorSystem, connectionFactory).withBufferSize(10).withQueue("test")
    )

  val webSocketFlow: Flow[ws.Message, ws.Message, Future[WebSocketUpgradeResponse]] = // (2)
    Http().webSocketClientFlow(WebSocketRequest("ws://localhost:8080/webSocket/ping"))

  val ((runningSource, wsUpgradeResponse), streamCompletion): ((JmsConsumerControl, Future[WebSocketUpgradeResponse]), Future[Done]) =
                                                     // stream element type
    jmsSource                                        //: String
      .map(ws.TextMessage(_))                        //: ws.TextMessage                  (3)
      .viaMat(webSocketFlow)(Keep.both)              //: ws.TextMessage                  (4)
      .mapAsync(1)(wsMessageToString)                //: String                          (5)
      .map("client received: " + _)                  //: String                          (6)
      .toMat(Sink.foreach(println))(Keep.both)       //                                  (7)
      .run()
  // #sample
  // format: on

  wsUpgradeResponse
    .map { upgrade =>
      if (upgrade.response.status == StatusCodes.SwitchingProtocols) {
        "WebSocket established"
      } else {
        throw new RuntimeException(s"Connection failed: ${upgrade.response.status}")
      }
    }
    .onComplete(println)

  wait(5.seconds)
  runningSource.shutdown()

  for {
    _ <- streamCompletion
    _ <- actorSystem.terminate()
    _ <- WebServer.stop()
    _ <- ActiveMqBroker.stop()
  } ()

  // #sample

  /**
   * Convert potentially chunked WebSocket Message to a string.
   */
  def wsMessageToString: ws.Message => Future[String] = {
    case message: ws.TextMessage.Strict =>
      Future.successful(message.text)

    case message: ws.TextMessage.Streamed =>
      val seq = message.textStream.runWith(Sink.seq)
      seq.map(seq => seq.mkString)

    case message =>
      Future.successful(message.toString)
  }
  // #sample

}

File: doc-examples/src/main/java/jms/javasamples/JmsToWebSocket.java
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <http://www.lightbend.com>
 */

package jms.javasamples;

// #sample

import akka.Done;
import akka.actor.ActorSystem;
import akka.http.javadsl.Http;
import akka.http.javadsl.model.StatusCodes;
import akka.http.javadsl.model.ws.Message;
import akka.http.javadsl.model.ws.TextMessage;
import akka.http.javadsl.model.ws.WebSocketRequest;
import akka.http.javadsl.model.ws.WebSocketUpgradeResponse;
import akka.japi.Pair;
import akka.stream.ActorMaterializer;
import akka.stream.Materializer;
import akka.stream.alpakka.jms.JmsConsumerSettings;
import akka.stream.alpakka.jms.JmsProducerSettings;
import akka.stream.alpakka.jms.javadsl.JmsConsumer;
import akka.stream.alpakka.jms.javadsl.JmsConsumerControl;
import akka.stream.alpakka.jms.javadsl.JmsProducer;
import akka.stream.javadsl.Flow;
import akka.stream.javadsl.Keep;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;
import playground.ActiveMqBroker;
import playground.WebServer;
import scala.concurrent.ExecutionContext;

import javax.jms.ConnectionFactory;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

// #sample

public class JmsToWebSocket {

  public static void main(String[] args) throws Exception {
    JmsToWebSocket me = new JmsToWebSocket();
    me.run();
  }

  private final ActorSystem system = ActorSystem.create();
  private final Materializer materializer = ActorMaterializer.create(system);
  private final ExecutionContext ec = system.dispatcher();

  private void enqueue(ConnectionFactory connectionFactory, String... msgs) {
    Sink<String, ?> jmsSink =
        JmsProducer.textSink(
            JmsProducerSettings.create(system, connectionFactory).withQueue("test"));
    Source.from(Arrays.asList(msgs)).runWith(jmsSink, materializer);
  }

  private void run() throws Exception {
    ActiveMqBroker activeMqBroker = new ActiveMqBroker();
    activeMqBroker.start();

    WebServer webserver = new WebServer();
    webserver.start("localhost", 8080);

    ConnectionFactory connectionFactory = activeMqBroker.createConnectionFactory();
    enqueue(connectionFactory, "a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k");
    // #sample

    final Http http = Http.get(system);

    Source<String, JmsConsumerControl> jmsSource = // (1)
        JmsConsumer.textSource(
            JmsConsumerSettings.create(system, connectionFactory)
                .withBufferSize(10)
                .withQueue("test"));

    Flow<Message, Message, CompletionStage<WebSocketUpgradeResponse>> webSocketFlow = // (2)
        http.webSocketClientFlow(WebSocketRequest.create("ws://localhost:8080/webSocket/ping"));

    int parallelism = 4;
    Pair<Pair<JmsConsumerControl, CompletionStage<WebSocketUpgradeResponse>>, CompletionStage<Done>>
        pair =
            jmsSource // : String
                .map(
                    s -> {
                      Message msg = TextMessage.create(s);
                      return msg;
                    }) // : Message           (3)
                .viaMat(webSocketFlow, Keep.both()) // : Message           (4)
                .mapAsync(parallelism, this::wsMessageToString) // : String            (5)
                .map(s -> "client received: " + s) // : String            (6)
                .toMat(Sink.foreach(System.out::println), Keep.both()) //                    (7)
                .run(materializer);
    // #sample
    JmsConsumerControl runningSource = pair.first().first();
    CompletionStage<WebSocketUpgradeResponse> wsUpgradeResponse = pair.first().second();
    CompletionStage<Done> streamCompletion = pair.second();

    wsUpgradeResponse
        .thenApply(
            upgrade -> {
              if (upgrade.response().status() == StatusCodes.SWITCHING_PROTOCOLS) {
                return "WebSocket established";
              } else {
                throw new RuntimeException("Connection failed: " + upgrade.response().status());
              }
            })
        .thenAccept(System.out::println);

    Thread.sleep(2 * 1000);
    runningSource.shutdown();
    streamCompletion.thenAccept(res -> system.terminate());
    system
        .getWhenTerminated()
        .thenAccept(
            t -> {
              webserver.stop();
              activeMqBroker.stop(ec);
            });
  }

  // #sample

  /** Convert potentially chunked WebSocket Message to a string. */
  private CompletionStage<String> wsMessageToString(Message msg) {
    if (msg.isText()) {
      TextMessage tMsg = msg.asTextMessage();
      if (tMsg.isStrict()) {
        return CompletableFuture.completedFuture(tMsg.getStrictText());
      } else {
        CompletionStage<List<String>> strings =
            tMsg.getStreamedText().runWith(Sink.seq(), materializer);
        return strings.thenApply(list -> String.join("", list));
      }
    } else {
      return CompletableFuture.completedFuture(msg.toString());
    }
  }
  // #sample

}
